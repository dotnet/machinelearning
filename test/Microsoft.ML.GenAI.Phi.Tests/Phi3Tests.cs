// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System.Text;
using System.Text.Json;
using ApprovalTests;
using ApprovalTests.Namers;
using ApprovalTests.Reporters;
using FluentAssertions;
using Microsoft.ML.GenAI.Core.Extension;
using Microsoft.ML.TestFramework;
using TorchSharp;
using Xunit;
using Xunit.Abstractions;
using static TorchSharp.torch;

namespace Microsoft.ML.GenAI.Phi.Tests;

public class Phi3Tests : BaseTestClass
{
    public Phi3Tests(ITestOutputHelper output) : base(output)
    {
        torch.set_default_device("meta");
    }

    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("Approvals")]
    public void Phi3Mini4KShapeTest()
    {
        var modelWeightFolder = "C:\\Users\\xiaoyuz\\source\\repos\\Phi-3-mini-4k-instruct";
        var config = Path.Join(modelWeightFolder, "config.json");
        var modelConfig = JsonSerializer.Deserialize<Phi3Config>(File.ReadAllText(config)) ?? throw new ArgumentNullException(nameof(config));
        var model = new Phi3ForCasualLM(modelConfig);
        var stateDictStr = model.PeekShape();
        Approvals.Verify(stateDictStr);
    }

    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("Approvals")]
    public void Phi3Medium4KShapeTest()
    {
        var modelWeightFolder = "C:\\Users\\xiaoyuz\\source\\repos\\Phi-3-medium-4k-instruct";
        var config = Path.Join(modelWeightFolder, "config.json");
        var modelConfig = JsonSerializer.Deserialize<Phi3Config>(File.ReadAllText(config)) ?? throw new ArgumentNullException(nameof(config));
        var model = new Phi3ForCasualLM(modelConfig);
        var stateDictStr = model.PeekShape();
        Approvals.Verify(stateDictStr);
    }


    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("Approvals")]
    public void Phi3Medium128KShapeTest()
    {
        var modelWeightFolder = "C:\\Users\\xiaoyuz\\source\\repos\\Phi-3-medium-128k-instruct";
        var config = Path.Join(modelWeightFolder, "config.json");
        var modelConfig = JsonSerializer.Deserialize<Phi3Config>(File.ReadAllText(config)) ?? throw new ArgumentNullException(nameof(config));
        var model = new Phi3ForCasualLM(modelConfig);
        var stateDictStr = model.PeekShape();
        Approvals.Verify(stateDictStr);
    }

    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("Approvals")]
    public void Phi3Mini128KShapeTest()
    {
        var modelWeightFolder = "C:\\Users\\xiaoyuz\\source\\repos\\Phi-3-mini-128k-instruct";
        var config = Path.Join(modelWeightFolder, "config.json");
        var modelConfig = JsonSerializer.Deserialize<Phi3Config>(File.ReadAllText(config)) ?? throw new ArgumentNullException(nameof(config));
        var model = new Phi3ForCasualLM(modelConfig);
        var stateDictStr = model.PeekShape();
        Approvals.Verify(stateDictStr);
    }

    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("Approvals")]
    public void Phi3Mini128KLayerSizeTest()
    {
        var dtype = ScalarType.BFloat16;
        var modelWeightFolder = "C:\\Users\\xiaoyuz\\source\\repos\\Phi-3-mini-128k-instruct";
        var config = Path.Join(modelWeightFolder, "config.json");
        var modelConfig = JsonSerializer.Deserialize<Phi3Config>(File.ReadAllText(config)) ?? throw new ArgumentNullException(nameof(config));
        modelConfig.DType = dtype;
        var model = new Phi3ForCasualLM(modelConfig);

        var size = model.GetSizeForEachDynamicLayerInBytes();
        // convert size to MB
        var sizeInMB = size.ToDictionary(x => x.Key, x => x.Value * 1.0f / 1024 / 1024);

        var json = JsonSerializer.Serialize(sizeInMB, new JsonSerializerOptions { WriteIndented = true });
        Approvals.Verify(json);
    }

    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("Approvals")]
    public void TokenizerTest()
    {
        var modelWeightFolder = "C:\\Users\\xiaoyuz\\source\\repos\\Phi-3-mini-4k-instruct";
        var tokenizer = LLama2Tokenizer.FromPretrained(modelWeightFolder);
        tokenizer.BosId.Should().Be(1);
        tokenizer.EosId.Should().Be(2);
        var messages = new string[]
        {
            "Can you provide ways to eat combinations of bananas and dragonfruits?",
            "Sure! Here are some ways to eat bananas and dragonfruits together: 1. Banana and dragonfruit smoothie: Blend bananas and dragonfruits together with some milk and honey. 2. Banana and dragonfruit salad: Mix sliced bananas and dragonfruits together with some lemon juice and honey.",
            "What about solving an 2x + 3 = 7 equation?",
            "\nCount to 3\n",
            "<|user|>\nCount to 3<|end|>\n<|assistant|>",
        };
        var sb = new StringBuilder();
        foreach (var message in messages)
        {
            var tokenized = tokenizer.Encode(message, true, false);
            var tokenizedStr = string.Join(", ", tokenized.Select(x => x.ToString()));

            sb.AppendLine(tokenizedStr);
        }
        Approvals.Verify(sb.ToString());
    }
}
