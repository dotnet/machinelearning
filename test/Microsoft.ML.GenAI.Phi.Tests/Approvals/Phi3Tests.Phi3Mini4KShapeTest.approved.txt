0: lm_head.weight shape: [32064, 3072]
1: model.embed_tokens.weight shape: [32064, 3072]
2: model.layers.0.input_layernorm.weight shape: [3072]
3: model.layers.0.mlp.down_proj.weight shape: [3072, 8192]
4: model.layers.0.mlp.gate_up_proj.weight shape: [16384, 3072]
5: model.layers.0.post_attention_layernorm.weight shape: [3072]
6: model.layers.0.self_attn.o_proj.weight shape: [3072, 3072]
7: model.layers.0.self_attn.qkv_proj.weight shape: [9216, 3072]
8: model.layers.1.input_layernorm.weight shape: [3072]
9: model.layers.1.mlp.down_proj.weight shape: [3072, 8192]
10: model.layers.1.mlp.gate_up_proj.weight shape: [16384, 3072]
11: model.layers.1.post_attention_layernorm.weight shape: [3072]
12: model.layers.1.self_attn.o_proj.weight shape: [3072, 3072]
13: model.layers.1.self_attn.qkv_proj.weight shape: [9216, 3072]
14: model.layers.10.input_layernorm.weight shape: [3072]
15: model.layers.10.mlp.down_proj.weight shape: [3072, 8192]
16: model.layers.10.mlp.gate_up_proj.weight shape: [16384, 3072]
17: model.layers.10.post_attention_layernorm.weight shape: [3072]
18: model.layers.10.self_attn.o_proj.weight shape: [3072, 3072]
19: model.layers.10.self_attn.qkv_proj.weight shape: [9216, 3072]
20: model.layers.11.input_layernorm.weight shape: [3072]
21: model.layers.11.mlp.down_proj.weight shape: [3072, 8192]
22: model.layers.11.mlp.gate_up_proj.weight shape: [16384, 3072]
23: model.layers.11.post_attention_layernorm.weight shape: [3072]
24: model.layers.11.self_attn.o_proj.weight shape: [3072, 3072]
25: model.layers.11.self_attn.qkv_proj.weight shape: [9216, 3072]
26: model.layers.12.input_layernorm.weight shape: [3072]
27: model.layers.12.mlp.down_proj.weight shape: [3072, 8192]
28: model.layers.12.mlp.gate_up_proj.weight shape: [16384, 3072]
29: model.layers.12.post_attention_layernorm.weight shape: [3072]
30: model.layers.12.self_attn.o_proj.weight shape: [3072, 3072]
31: model.layers.12.self_attn.qkv_proj.weight shape: [9216, 3072]
32: model.layers.13.input_layernorm.weight shape: [3072]
33: model.layers.13.mlp.down_proj.weight shape: [3072, 8192]
34: model.layers.13.mlp.gate_up_proj.weight shape: [16384, 3072]
35: model.layers.13.post_attention_layernorm.weight shape: [3072]
36: model.layers.13.self_attn.o_proj.weight shape: [3072, 3072]
37: model.layers.13.self_attn.qkv_proj.weight shape: [9216, 3072]
38: model.layers.14.input_layernorm.weight shape: [3072]
39: model.layers.14.mlp.down_proj.weight shape: [3072, 8192]
40: model.layers.14.mlp.gate_up_proj.weight shape: [16384, 3072]
41: model.layers.14.post_attention_layernorm.weight shape: [3072]
42: model.layers.14.self_attn.o_proj.weight shape: [3072, 3072]
43: model.layers.14.self_attn.qkv_proj.weight shape: [9216, 3072]
44: model.layers.15.input_layernorm.weight shape: [3072]
45: model.layers.15.mlp.down_proj.weight shape: [3072, 8192]
46: model.layers.15.mlp.gate_up_proj.weight shape: [16384, 3072]
47: model.layers.15.post_attention_layernorm.weight shape: [3072]
48: model.layers.15.self_attn.o_proj.weight shape: [3072, 3072]
49: model.layers.15.self_attn.qkv_proj.weight shape: [9216, 3072]
50: model.layers.16.input_layernorm.weight shape: [3072]
51: model.layers.16.mlp.down_proj.weight shape: [3072, 8192]
52: model.layers.16.mlp.gate_up_proj.weight shape: [16384, 3072]
53: model.layers.16.post_attention_layernorm.weight shape: [3072]
54: model.layers.16.self_attn.o_proj.weight shape: [3072, 3072]
55: model.layers.16.self_attn.qkv_proj.weight shape: [9216, 3072]
56: model.layers.17.input_layernorm.weight shape: [3072]
57: model.layers.17.mlp.down_proj.weight shape: [3072, 8192]
58: model.layers.17.mlp.gate_up_proj.weight shape: [16384, 3072]
59: model.layers.17.post_attention_layernorm.weight shape: [3072]
60: model.layers.17.self_attn.o_proj.weight shape: [3072, 3072]
61: model.layers.17.self_attn.qkv_proj.weight shape: [9216, 3072]
62: model.layers.18.input_layernorm.weight shape: [3072]
63: model.layers.18.mlp.down_proj.weight shape: [3072, 8192]
64: model.layers.18.mlp.gate_up_proj.weight shape: [16384, 3072]
65: model.layers.18.post_attention_layernorm.weight shape: [3072]
66: model.layers.18.self_attn.o_proj.weight shape: [3072, 3072]
67: model.layers.18.self_attn.qkv_proj.weight shape: [9216, 3072]
68: model.layers.19.input_layernorm.weight shape: [3072]
69: model.layers.19.mlp.down_proj.weight shape: [3072, 8192]
70: model.layers.19.mlp.gate_up_proj.weight shape: [16384, 3072]
71: model.layers.19.post_attention_layernorm.weight shape: [3072]
72: model.layers.19.self_attn.o_proj.weight shape: [3072, 3072]
73: model.layers.19.self_attn.qkv_proj.weight shape: [9216, 3072]
74: model.layers.2.input_layernorm.weight shape: [3072]
75: model.layers.2.mlp.down_proj.weight shape: [3072, 8192]
76: model.layers.2.mlp.gate_up_proj.weight shape: [16384, 3072]
77: model.layers.2.post_attention_layernorm.weight shape: [3072]
78: model.layers.2.self_attn.o_proj.weight shape: [3072, 3072]
79: model.layers.2.self_attn.qkv_proj.weight shape: [9216, 3072]
80: model.layers.20.input_layernorm.weight shape: [3072]
81: model.layers.20.mlp.down_proj.weight shape: [3072, 8192]
82: model.layers.20.mlp.gate_up_proj.weight shape: [16384, 3072]
83: model.layers.20.post_attention_layernorm.weight shape: [3072]
84: model.layers.20.self_attn.o_proj.weight shape: [3072, 3072]
85: model.layers.20.self_attn.qkv_proj.weight shape: [9216, 3072]
86: model.layers.21.input_layernorm.weight shape: [3072]
87: model.layers.21.mlp.down_proj.weight shape: [3072, 8192]
88: model.layers.21.mlp.gate_up_proj.weight shape: [16384, 3072]
89: model.layers.21.post_attention_layernorm.weight shape: [3072]
90: model.layers.21.self_attn.o_proj.weight shape: [3072, 3072]
91: model.layers.21.self_attn.qkv_proj.weight shape: [9216, 3072]
92: model.layers.22.input_layernorm.weight shape: [3072]
93: model.layers.22.mlp.down_proj.weight shape: [3072, 8192]
94: model.layers.22.mlp.gate_up_proj.weight shape: [16384, 3072]
95: model.layers.22.post_attention_layernorm.weight shape: [3072]
96: model.layers.22.self_attn.o_proj.weight shape: [3072, 3072]
97: model.layers.22.self_attn.qkv_proj.weight shape: [9216, 3072]
98: model.layers.23.input_layernorm.weight shape: [3072]
99: model.layers.23.mlp.down_proj.weight shape: [3072, 8192]
100: model.layers.23.mlp.gate_up_proj.weight shape: [16384, 3072]
101: model.layers.23.post_attention_layernorm.weight shape: [3072]
102: model.layers.23.self_attn.o_proj.weight shape: [3072, 3072]
103: model.layers.23.self_attn.qkv_proj.weight shape: [9216, 3072]
104: model.layers.24.input_layernorm.weight shape: [3072]
105: model.layers.24.mlp.down_proj.weight shape: [3072, 8192]
106: model.layers.24.mlp.gate_up_proj.weight shape: [16384, 3072]
107: model.layers.24.post_attention_layernorm.weight shape: [3072]
108: model.layers.24.self_attn.o_proj.weight shape: [3072, 3072]
109: model.layers.24.self_attn.qkv_proj.weight shape: [9216, 3072]
110: model.layers.25.input_layernorm.weight shape: [3072]
111: model.layers.25.mlp.down_proj.weight shape: [3072, 8192]
112: model.layers.25.mlp.gate_up_proj.weight shape: [16384, 3072]
113: model.layers.25.post_attention_layernorm.weight shape: [3072]
114: model.layers.25.self_attn.o_proj.weight shape: [3072, 3072]
115: model.layers.25.self_attn.qkv_proj.weight shape: [9216, 3072]
116: model.layers.26.input_layernorm.weight shape: [3072]
117: model.layers.26.mlp.down_proj.weight shape: [3072, 8192]
118: model.layers.26.mlp.gate_up_proj.weight shape: [16384, 3072]
119: model.layers.26.post_attention_layernorm.weight shape: [3072]
120: model.layers.26.self_attn.o_proj.weight shape: [3072, 3072]
121: model.layers.26.self_attn.qkv_proj.weight shape: [9216, 3072]
122: model.layers.27.input_layernorm.weight shape: [3072]
123: model.layers.27.mlp.down_proj.weight shape: [3072, 8192]
124: model.layers.27.mlp.gate_up_proj.weight shape: [16384, 3072]
125: model.layers.27.post_attention_layernorm.weight shape: [3072]
126: model.layers.27.self_attn.o_proj.weight shape: [3072, 3072]
127: model.layers.27.self_attn.qkv_proj.weight shape: [9216, 3072]
128: model.layers.28.input_layernorm.weight shape: [3072]
129: model.layers.28.mlp.down_proj.weight shape: [3072, 8192]
130: model.layers.28.mlp.gate_up_proj.weight shape: [16384, 3072]
131: model.layers.28.post_attention_layernorm.weight shape: [3072]
132: model.layers.28.self_attn.o_proj.weight shape: [3072, 3072]
133: model.layers.28.self_attn.qkv_proj.weight shape: [9216, 3072]
134: model.layers.29.input_layernorm.weight shape: [3072]
135: model.layers.29.mlp.down_proj.weight shape: [3072, 8192]
136: model.layers.29.mlp.gate_up_proj.weight shape: [16384, 3072]
137: model.layers.29.post_attention_layernorm.weight shape: [3072]
138: model.layers.29.self_attn.o_proj.weight shape: [3072, 3072]
139: model.layers.29.self_attn.qkv_proj.weight shape: [9216, 3072]
140: model.layers.3.input_layernorm.weight shape: [3072]
141: model.layers.3.mlp.down_proj.weight shape: [3072, 8192]
142: model.layers.3.mlp.gate_up_proj.weight shape: [16384, 3072]
143: model.layers.3.post_attention_layernorm.weight shape: [3072]
144: model.layers.3.self_attn.o_proj.weight shape: [3072, 3072]
145: model.layers.3.self_attn.qkv_proj.weight shape: [9216, 3072]
146: model.layers.30.input_layernorm.weight shape: [3072]
147: model.layers.30.mlp.down_proj.weight shape: [3072, 8192]
148: model.layers.30.mlp.gate_up_proj.weight shape: [16384, 3072]
149: model.layers.30.post_attention_layernorm.weight shape: [3072]
150: model.layers.30.self_attn.o_proj.weight shape: [3072, 3072]
151: model.layers.30.self_attn.qkv_proj.weight shape: [9216, 3072]
152: model.layers.31.input_layernorm.weight shape: [3072]
153: model.layers.31.mlp.down_proj.weight shape: [3072, 8192]
154: model.layers.31.mlp.gate_up_proj.weight shape: [16384, 3072]
155: model.layers.31.post_attention_layernorm.weight shape: [3072]
156: model.layers.31.self_attn.o_proj.weight shape: [3072, 3072]
157: model.layers.31.self_attn.qkv_proj.weight shape: [9216, 3072]
158: model.layers.4.input_layernorm.weight shape: [3072]
159: model.layers.4.mlp.down_proj.weight shape: [3072, 8192]
160: model.layers.4.mlp.gate_up_proj.weight shape: [16384, 3072]
161: model.layers.4.post_attention_layernorm.weight shape: [3072]
162: model.layers.4.self_attn.o_proj.weight shape: [3072, 3072]
163: model.layers.4.self_attn.qkv_proj.weight shape: [9216, 3072]
164: model.layers.5.input_layernorm.weight shape: [3072]
165: model.layers.5.mlp.down_proj.weight shape: [3072, 8192]
166: model.layers.5.mlp.gate_up_proj.weight shape: [16384, 3072]
167: model.layers.5.post_attention_layernorm.weight shape: [3072]
168: model.layers.5.self_attn.o_proj.weight shape: [3072, 3072]
169: model.layers.5.self_attn.qkv_proj.weight shape: [9216, 3072]
170: model.layers.6.input_layernorm.weight shape: [3072]
171: model.layers.6.mlp.down_proj.weight shape: [3072, 8192]
172: model.layers.6.mlp.gate_up_proj.weight shape: [16384, 3072]
173: model.layers.6.post_attention_layernorm.weight shape: [3072]
174: model.layers.6.self_attn.o_proj.weight shape: [3072, 3072]
175: model.layers.6.self_attn.qkv_proj.weight shape: [9216, 3072]
176: model.layers.7.input_layernorm.weight shape: [3072]
177: model.layers.7.mlp.down_proj.weight shape: [3072, 8192]
178: model.layers.7.mlp.gate_up_proj.weight shape: [16384, 3072]
179: model.layers.7.post_attention_layernorm.weight shape: [3072]
180: model.layers.7.self_attn.o_proj.weight shape: [3072, 3072]
181: model.layers.7.self_attn.qkv_proj.weight shape: [9216, 3072]
182: model.layers.8.input_layernorm.weight shape: [3072]
183: model.layers.8.mlp.down_proj.weight shape: [3072, 8192]
184: model.layers.8.mlp.gate_up_proj.weight shape: [16384, 3072]
185: model.layers.8.post_attention_layernorm.weight shape: [3072]
186: model.layers.8.self_attn.o_proj.weight shape: [3072, 3072]
187: model.layers.8.self_attn.qkv_proj.weight shape: [9216, 3072]
188: model.layers.9.input_layernorm.weight shape: [3072]
189: model.layers.9.mlp.down_proj.weight shape: [3072, 8192]
190: model.layers.9.mlp.gate_up_proj.weight shape: [16384, 3072]
191: model.layers.9.post_attention_layernorm.weight shape: [3072]
192: model.layers.9.self_attn.o_proj.weight shape: [3072, 3072]
193: model.layers.9.self_attn.qkv_proj.weight shape: [9216, 3072]
194: model.norm.weight shape: [3072]
